client/src/main/scala/drt/client/services/handlers/CrunchStateUpdatesHandler.scala

```diff
@@ -0,0 +1,119 @@
+package drt.client.services.handlers
+
+import diode._
+import diode.data._
+import diode.Implicits.runAfterImpl
+import drt.client.actions.Actions._
+import drt.client.logger._
+import drt.client.services._
+import drt.shared.{ApiFlightWithSplits, CrunchApi}
+import drt.shared.CrunchApi._
+import org.scalajs.dom
+import upickle.default.read
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
+
+class CrunchStateUpdatesHandler[M](getCurrentViewMode: () => ViewMode,
+                                   modelRW: ModelRW[M, (Pot[CrunchState], MillisSinceEpoch)]) extends LoggingActionHandler(modelRW) {
+  val liveRequestFrequency: FiniteDuration = 2 seconds
+  val forecastRequestFrequency: FiniteDuration = 15 seconds
+
+  protected def handle: PartialFunction[Any, ActionResult[M]] = {
+    case GetCrunchStateUpdates(viewMode) =>
+      val (_, lastUpdateMillis) = modelRW.value
+      val startMillis = startMillisFromView
+      val endMillis = startMillis + (1000 * 60 * 60 * 36)
+      val updateRequestFuture = DrtApi.get(s"crunch?start=$startMillis&end=$endMillis&since=$lastUpdateMillis")
+
+      val eventualAction = processUpdatesRequest(viewMode, updateRequestFuture)
+
+      effectOnly(Effect(eventualAction))
+
+    case UpdateCrunchStateFromUpdates(viewMode, _) if viewMode.dayStart.millisSinceEpoch != startMillisFromView =>
+      log.warn(s"Received old crunch updates response (${viewMode.dayStart.millisSinceEpoch} != $startMillisFromView)")
+      noChange
+
+    case UpdateCrunchStateFromUpdates(viewMode, crunchUpdates) =>
+      val (Ready(existingState), _) = modelRW.value
```

Comment:
This should probably do a match to be safer.

Reply:
Good catch!